Properly handle both v1 and v2 span-ids.

csilvers opened the pull request:
The SPAN_ID is a decimal number in v1 of the trace-context API, but a (16-digit) hexadecimal number in v2:
   https://cloud.google.com/python/docs/reference/cloudtrace/1.2.0/google.cloud.trace_v1.types.TraceSpan
   https://cloud.google.com/python/docs/reference/cloudtrace/1.2.0/google.cloud.trace_v2.types.Span

Stackdriver uses only v2 these days:
   https://issuetracker.google.com/issues/338634230?pli=1

Unfortunately, the X-Cloud-Trace-Context may still be using the v1 API.  We try to detect that situation and convert it to a v2 compatible trace-id by formatting the number in hex.

We cannot perfectly detect whether we're v1 or v2, because there are some decimal numbers that look like 16-digit hex numbers.  But not very many; we'd expect the logic below to make a mistake about 1 in every 2,000 log entries.  In such cases, we will use the wrong span-id, making it harder to associate log-messages with the requestlog that spawned it.

Issue: https://console.cloud.google.com/support/cases/detail/v2/53379415?project=khan-academy

Test plan:
Fingers crossed, it can't be any worse than what we have now!

csilvers committed:
Properly handle both v1 and v2 span-ids.

The SPAN_ID is a decimal number in v1 of the trace-context API,
but a (16-digit) hexadecimal number in v2:
   https://cloud.google.com/python/docs/reference/cloudtrace/1.2.0/google.cloud.trace_v1.types.TraceSpan
   https://cloud.google.com/python/docs/reference/cloudtrace/1.2.0/google.cloud.trace_v2.types.Span

Stackdriver uses only v2 these days:
   https://issuetracker.google.com/issues/338634230?pli=1

Unfortunately, the X-Cloud-Trace-Context may still be using the
v1 API.  We try to detect that situation and convert it to a v2
compatible trace-id by formatting the number in hex.

We cannot perfectly detect whether we're v1 or v2, because
there are some decimal numbers that look like 16-digit hex
numbers.  But not very many; we'd expect the logic below to
make a mistake about 1 in every 2,000 log entries.  In such
cases, we will use the wrong span-id, making it harder to
associate log-messages with the requestlog that spawned it.

Issue: https://console.cloud.google.com/support/cases/detail/v2/53379415?project=khan-academy

Test plan:
Fingers crossed, it can't be any worse than what we have now!

andyfusniak commented:
Thanks for the pull request. I only just noticed this pull request and didn't get a notification email from GitHub.

I haven't used Google Cloud Run for a while now, so I'm going off of the documentation links you sent.

I was hoping there would be a way to tell the v1 and v2 requests apart, but there seems to be no reliable headers to differentiate the versions. Also, both v1 and v2 use the same format for trace IDs.

I did wonder if having an option in the package to allow you to manually indicate the version might be helpful. But I can't be certain that Google wouldn't mix v1 and v2 requests within the same project, or services or components within the same project might be using different versions. This might cause more problems than it's worth. It could serve as a hint to swing the answer in the times when a span ID with 16 decimal digits comes up, but that might be overengineered for a problem that only affects about 0.05% of requests.

That only leaves the examination of the span IDs and you have done. I wasn't able to come up with any better solution to the problem, so I'll be merging the pull request in a moment.

csilvers commented:
Hi @andyfusniak thanks for taking a look at this! I've been talking to Google and that said they are going to move entirely to v2 over the next few months, so it may be that this isn't needed anymore. It might be appropriate to revert this pr when that happens. I'll note in this pr when they tell me it's happened!

andyfusniak commented:
That's good to know. When Google move entirely to v2 I'll tidy it up and roll out another version. I've put this pr under v0.2.1 which also includes a Go dependancy security update for a standard library internal package.

The go.mod file still requires an outdated version of logrus (around v1.4.x I think, and we're now at v1.9.x) so I ought to bring this up to date. This package was knocked up a while back and contains no unit test, unfortunately, so I didn't feel confident to bump it up without being able to test it.

Let me know if you get to any issues, unless you're happy using your forked copy.

Regards,
Andy

csilvers commented:
So I talked to Google and discovered that they are never going to move to
v1, for backwards-compatibility reasons.  I cleaned up the code based on
this, and fixed a bug, and have a new pull request out for that:
https://github.com/andyfusniak/stackdriver-gae-logrus-plugin/pull/3.  After
that's in, I think there's not anything more to do here.

craig

On Tue, Oct 29, 2024 at 3:49 PM Andy Fusniak ***@***.***>
wrote:

> That's good to know. When Google move entirely to v2 I'll tidy it up and
> roll out another version. I've put this pr under v0.2.1 which also includes
> a Go dependancy security update for a standard library internal package.
>
> The go.mod file still requires an outdated version of logrus (around
> v1.4.x I think, and we're now at v1.9.x) so I ought to bring this up to
> date. This package was knocked up a while back and contains no unit test,
> unfortunately, so I didn't feel confident to bump it up without being able
> to test it.
>
> Let me know if you get to any issues, unless you're happy using your
> forked copy.
>
> Regards,
> Andy
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/andyfusniak/stackdriver-gae-logrus-plugin/pull/2#issuecomment-2445465882>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALA2Z3RR4OJRAZLKMODFRLZ6AGH3AVCNFSM6AAAAABOQ4KBH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBVGQ3DKOBYGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
